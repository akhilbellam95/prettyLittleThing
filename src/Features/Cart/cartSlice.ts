import { createSlice } from '@reduxjs/toolkit';

interface CartItem {
  id: number;
  quantity: number;
};

export interface CartState {
  cartItems: CartItem[],
  total: number
}

export const cartInitialState: CartState = {
  cartItems: [] as CartItem[],
  total: 0,
};

export const cartSlice = createSlice({
  name: 'cart',
  initialState: cartInitialState,
  reducers: {
    addToCart: (state, action) => {
      const { id } = action.payload;
      const existingItem = state.cartItems.find((item) => item.id === id);
      let total = state.total;
      if (existingItem) {
        existingItem.quantity++;
      } else {
        state.cartItems.push({ id, quantity: 1 });
      }
      total += action.payload.price;
      state.total = parseFloat(total.toFixed(2));
    },
    removeFromCart: (state, action) => {
      const itemIndex = state.cartItems.findIndex((el) => el.id === action.payload.id);
      if (itemIndex !== -1) {
        const item = state.cartItems[itemIndex];
        let total = state.total;
        if (item.quantity === 1) {
          state.cartItems.splice(itemIndex, 1);
        } else {
          item.quantity--;
        }
        total -= action.payload.price;
        state.total = parseFloat(total.toFixed(2));
      }
    },
    updateTotal: (state, action) => {
      state.total = action.payload;
    },
  },
});

export const { addToCart, removeFromCart, updateTotal } = cartSlice.actions;

export default cartSlice.reducer;
