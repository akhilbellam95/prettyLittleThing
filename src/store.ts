// eslint-disable-next-line prettier/prettier
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from './Features/Cart/cartSlice';
import productReducer from './Features/Product/productSlice';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    product: productReducer,
  },
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;

export type RootState = ReturnType<typeof store.getState>;
