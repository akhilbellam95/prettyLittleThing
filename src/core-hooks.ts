import { configureStore, createStore } from "@reduxjs/toolkit";
import { useSelector } from "react-redux";
import { TypedUseSelectorHook, useDispatch } from "react-redux";
import { CartState } from "./Features/Cart/cartSlice";
import { ProductsState } from "./Features/Product/productSlice";

export interface ApplicationState {
  cart: CartState;
  product: ProductsState;
}

export const useAppDispatch = () =>
  useDispatch<ReturnType<typeof configureStore>['dispatch']>();

export const useAppSelector: TypedUseSelectorHook<ApplicationState> =
  useSelector;