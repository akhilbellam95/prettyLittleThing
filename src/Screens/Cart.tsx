import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import CartItems from '../Components/CartItems';
import { ApplicationState, useAppSelector } from '../core-hooks';

const Cart = () => {
  const { cartItems } = useAppSelector((state: ApplicationState) => state.cart);

  return (
    <View style={styles.container}>
      {cartItems.length > 0 ? <CartItems /> : <Text>Cart is empty</Text>}
    </View>
  );
};

export default Cart;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
