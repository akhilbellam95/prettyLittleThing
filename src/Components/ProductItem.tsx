import { Pressable, Text, View } from 'react-native';
import React from 'react';
import { addToCart, removeFromCart } from '../Features/Cart/cartSlice';
import { RootState } from '../store';
import AddRemoveItem from './AddRemoveItem';
import FastImage from 'react-native-fast-image';
import { Product, styles } from './Products';
import { ApplicationState, useAppDispatch, useAppSelector } from '../core-hooks';

export const ProductItem = ({ product }: { product: Product }): JSX.Element => {
  const { cartItems } = useAppSelector((state: ApplicationState) => state.cart);

  const itemInCart = cartItems.find((el): boolean => el.id === product.id);
  const itemIndex = cartItems.findIndex((el): boolean => el.id === product.id);
  const dispatch = useAppDispatch();
  return (
    <View style={styles.card}>
      <View style={styles.alignCenter}>
        <FastImage
          source={{ uri: product.img, priority: FastImage.priority.normal }}
          style={styles.productImage}
          resizeMode={FastImage.resizeMode.contain}
        />
      </View>

      <View style={styles.productDetails}>
        <View style={styles.row}>
          <Text style={styles.label}>Name: </Text>
          <Text style={styles.value}>{product.name}</Text>
        </View>

        <View style={styles.row}>
          <Text style={styles.label}>Price: </Text>
          <Text style={styles.value}>{product.price}</Text>
        </View>

        <View style={styles.row}>
          <Text style={styles.label}>Colour: </Text>
          <Text style={styles.value}>{product.colour}</Text>
        </View>
      </View>

      <View style={styles.seperator} />
      <View style={{ alignItems: 'flex-end' }}>
        {itemInCart ? (
          // Add & Remove from cart buttons with quantity
          <AddRemoveItem
            addPressed={() => dispatch(addToCart(product))}
            removePressed={() => dispatch(removeFromCart(product))}
            quantity={cartItems[itemIndex].quantity}
          />
        ) : (
          // Add to cart button
          <Pressable
            style={styles.button}
            testID='add-to-cart'
            onPress={() => dispatch(addToCart(product))}
          >
            <Text style={styles.textCenter}>Add to Cart</Text>
          </Pressable>
        )}
      </View>
    </View>
  );
};
