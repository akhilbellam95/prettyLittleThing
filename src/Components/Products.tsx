import { FlatList, StyleSheet, View } from 'react-native';
import React, { useEffect } from 'react';
import { getProducts } from '../Features/Product/productSlice';
import { ApplicationState, useAppDispatch, useAppSelector } from '../core-hooks';
import { ProductItem } from './ProductItem';

export interface Product {
  id: number;
  name: string;
  img: string;
  price: number;
  colour: string;
}

interface Products {
  products: Product[];
}

const Products = () => {
  const { products } = useAppSelector((state: ApplicationState) => state.product);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getProducts());
  }, []);

  return (
    <View>
      <FlatList data={products} renderItem={({ item }) => <ProductItem product={item} />} />
    </View>
  );
};

export default Products;

export const styles = StyleSheet.create({
  alignCenter: {
    alignItems: 'center',
  },
  button: {
    backgroundColor: 'yellow',
    width: 120,
    height: 40,
    justifyContent: 'center',
    position: 'relative',
    paddingHorizontal: 20,
    borderWidth: 1,
    borderColor: 'black',
  },
  card: {
    padding: 20,
    borderWidth: 1,
    marginHorizontal: 20,
    marginVertical: 10,
  },
  productDetails: {
    borderTopWidth: 1,
    borderTopColor: 'black',
    marginTop: 20,
    paddingTop: 15,
  },
  productImage: {
    width: 280,
    height: 500,
  },
  label: {
    width: 60,
    marginBottom: 5,
  },
  row: {
    flexDirection: 'row',
  },
  seperator: {
    height: 20,
  },
  textCenter: {
    textAlign: 'center',
  },
  value: {
    width: '75%',
  },
});
