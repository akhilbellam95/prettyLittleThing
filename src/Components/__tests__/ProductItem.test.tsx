import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import Products from '../Products';
import { useAppDispatch, useAppSelector } from '../../core-hooks';
import { mockCartItems, mockProducts } from '../../mocks/data';
import { act } from 'react-test-renderer';
import { store } from '../../store';
import { addToCart } from '../../Features/Cart/cartSlice';
import { ProductItem } from '../ProductItem';

jest.mock('../../core-hooks', () => ({
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

describe('ProductItem component testing', () => {
  // let store;
  const mockDispatch = jest.fn();
  const mockSelector = jest.fn();
  const tree = (
    <Provider store={store}>
      <ProductItem product={mockProducts[0]} />
    </Provider>
  );

  beforeAll(() => {
    (useAppSelector as jest.Mock).mockImplementation(mockSelector);
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  beforeEach(() => {
    mockSelector.mockReturnValue({
      product: mockProducts,
      cartItems: [],
    });
  });

  it('renders correctly', () => {
    const snap: unknown = render(tree).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it('should add product to cart', () => {
    const product = { id: 1 };
    store.dispatch(addToCart(product));
    const state = store.getState()['cart'];
    expect(state.cartItems).toHaveLength(1);
    expect(state.cartItems[0]).toEqual({ ...product, quantity: 1 });

    const snap: unknown = render(tree).toJSON();
    expect(snap).toMatchSnapshot();
  });

  test('should call when user press Add to cart button', async () => {
    const { getByTestId } = render(tree);

    await act(() => {
      fireEvent.press(getByTestId('add-to-cart'));
    });
  });

  test('should call when user press add button', async () => {
    mockSelector.mockReturnValue({
      cartItems: mockCartItems,
    });
    const { getByTestId } = render(tree);

    await act(() => {
      fireEvent.press(getByTestId('add-btn'));
    });
  });

  test('should call when user press remove button', async () => {
    mockSelector.mockReturnValue({
      cartItems: mockCartItems,
    });
    const { getByTestId } = render(tree);

    await act(() => {
      fireEvent.press(getByTestId('remove-btn'));
    });
  });
});
