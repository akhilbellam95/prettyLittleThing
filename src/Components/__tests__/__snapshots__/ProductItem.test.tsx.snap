// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem component testing renders correctly 1`] = `
<View
  style={
    {
      "borderWidth": 1,
      "marginHorizontal": 20,
      "marginVertical": 10,
      "padding": 20,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "height": 500,
            "width": 280,
          },
        ]
      }
    >
      <FastImageView
        defaultSource={null}
        resizeMode="contain"
        source={
          {
            "priority": "normal",
            "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
          }
        }
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "borderTopColor": "black",
        "borderTopWidth": 1,
        "marginTop": 20,
        "paddingTop": 15,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Name: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        Black Sheet Strappy Textured Glitter Bodycon Dress
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Price: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        10
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Colour: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        Black
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 20,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "flex-end",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "yellow",
          "borderColor": "black",
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "position": "relative",
          "width": 120,
        }
      }
      testID="add-to-cart"
    >
      <Text
        style={
          {
            "textAlign": "center",
          }
        }
      >
        Add to Cart
      </Text>
    </View>
  </View>
</View>
`;

exports[`ProductItem component testing should add product to cart 1`] = `
<View
  style={
    {
      "borderWidth": 1,
      "marginHorizontal": 20,
      "marginVertical": 10,
      "padding": 20,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "height": 500,
            "width": 280,
          },
        ]
      }
    >
      <FastImageView
        defaultSource={null}
        resizeMode="contain"
        source={
          {
            "priority": "normal",
            "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
          }
        }
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "borderTopColor": "black",
        "borderTopWidth": 1,
        "marginTop": 20,
        "paddingTop": 15,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Name: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        Black Sheet Strappy Textured Glitter Bodycon Dress
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Price: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        10
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "marginBottom": 5,
            "width": 60,
          }
        }
      >
        Colour: 
      </Text>
      <Text
        style={
          {
            "width": "75%",
          }
        }
      >
        Black
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 20,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "flex-end",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "yellow",
          "borderColor": "black",
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 20,
          "position": "relative",
          "width": 120,
        }
      }
      testID="add-to-cart"
    >
      <Text
        style={
          {
            "textAlign": "center",
          }
        }
      >
        Add to Cart
      </Text>
    </View>
  </View>
</View>
`;
